robot/impl/resource_manager_test.go

```diff
@@ -1583,6 +1584,137 @@ func TestReconfigure(t *testing.T) {
 	}()
 }
 
+func TestRemoteConnClosedOnReconfigure(t *testing.T) {
+	logger, observer := logging.NewObservedTestLogger(t)
+
+	ctx := context.Background()
+
+	fakeMotor := resource.Config{
+		Name:                "motor",
+		Model:               resource.DefaultModelFamily.WithModel("fake"),
+		API:                 motor.API,
+		ConvertedAttributes: &fakemotor.Config{},
+	}
+
+	fakeArm := resource.Config{
+		Name:                "arm",
+		Model:               resource.DefaultModelFamily.WithModel("fake"),
+		API:                 arm.API,
+		ConvertedAttributes: &fakearm.Config{},
+	}
+
+	remoteCfg1 := &config.Config{
+		Components: []resource.Config{fakeMotor},
+	}
+
+	remoteCfg2 := &config.Config{
+		Components: []resource.Config{fakeArm},
+	}
+
+	t.Run("remotes with same exact resources", func(t *testing.T) {
+		remote1 := setupLocalRobot(t, ctx, remoteCfg1, logger.Sublogger("remote1"))
+		remote2 := setupLocalRobot(t, ctx, remoteCfg1, logger.Sublogger("remote2"))
+
+		options1, _, addr1 := robottestutils.CreateBaseOptionsAndListener(t)
+		err := remote1.StartWeb(ctx, options1)
+		test.That(t, err, test.ShouldBeNil)
+
+		options2, _, addr2 := robottestutils.CreateBaseOptionsAndListener(t)
+		err = remote2.StartWeb(ctx, options2)
+		test.That(t, err, test.ShouldBeNil)
+
+		// Set up a local main robot which is connected to remote1
+		remoteConf := config.Remote{
+			Name:    "remote",
+			Address: addr1,
+		}
+
+		mainRobotCfg := &config.Config{
+			Remotes: []config.Remote{remoteConf},
+		}
+
+		// Make a copy of the main robot config as reconfigure will directly modify it
+		mainCfgCopy := *mainRobotCfg
+		mainRobot := setupLocalRobot(t, context.Background(), mainRobotCfg, logger.Sublogger("main"))
+
+		// Change address in config copy to reference address of the remote2, make a copy of the config, and reconfigure
+		mainCfgCopy.Remotes[0].Address = addr2
+		mainCfgCopy2 := mainCfgCopy
+		mainRobot.Reconfigure(context.Background(), &mainCfgCopy)
+
+		// Change address back to original remote, and reconfigure
+		mainCfgCopy2.Remotes[0].Address = addr1
+		mainRobot.Reconfigure(context.Background(), &mainCfgCopy2)
+
+		// Check that we were able to grab the motor from remote1 through the main robot and successfully call IsPowered()
+		remoteFromMain, ok := mainRobot.RemoteByName("remote")
+		test.That(t, ok, test.ShouldBeTrue)
+
+		res, err := motor.FromRobot(remoteFromMain, "motor")
+		test.That(t, err, test.ShouldBeNil)
+
+		moving, speed, err := res.IsPowered(context.Background(), nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, moving, test.ShouldBeFalse)
+		test.That(t, speed, test.ShouldEqual, 0.0)
+
+		// Also check that there are no error logs associated with the main robot trying to reconnect to remote2
```

Comment:
why would the main robot be trying to reconnect to remote2 if we aren't shutting remote2 down?

Reply:
right, I originally had that but removed it because it would fail without anyway, but doesn't make much sense without it

Reply:
what is "it" in this scenario? should we remove the assertion below then?

Reply:
"it" was calling close on remote 2 which I re-added. The test will work with or without it, but I chose to include it originally since the main indication of remote leaks was error logs related to reconnecting